packages/report-flaky-tests/lib/markdown.ts

```diff
@@ -0,0 +1,201 @@
+import * as core from '@actions/core';
+
+/**
+ * Internal dependencies
+ */
+import { stripAnsi } from './strip-ansi';
+import { TEST_RESULTS_LIST, TEST_RESULT, metaData } from './constants';
+import type { MetaData, FlakyTestResult } from './types';
+
+type ParsedTestResult = {
+	date: Date;
+	failedTimes: number;
+	runURL: string;
+	headBranch: string;
+	errorMessage?: string;
+};
+
+const TEST_LOG_REGEX = new RegExp(
+	`<time datetime="(?<date>.+)">.*</time>\\s*Test passed after (?<failedTimes>\\d) failed attempts? on <a href="(?<runURL>.+)"><code>(?<headBranch>.+)</code></a>\\.`
+);
+const TEST_FULL_REGEX = new RegExp(
+	`(?:<details>\\s*<summary>\\s*)?${ TEST_LOG_REGEX.source }(?:\\s*</summary>\\s*
+\`\`\`
+(?<errorMessage>[\\s\\S]*)
+\`\`\`
+\\s*</details>)?`
+);
+
+function renderIssueBody( {
+	meta,
+	testTitle,
+	testPath,
+	formattedTestResults,
+}: {
+	meta: MetaData;
+	testTitle: string;
+	testPath: string;
+	formattedTestResults: string;
+} ) {
+	return `${ metaData.render( meta ) }
+**Flaky test detected. This is an auto-generated issue by GitHub Actions. Please do NOT edit this manually.**
+
+## Test title
+${ testTitle }
+
+## Test path
+\`${ testPath }\`
+
+## Errors
+${ TEST_RESULTS_LIST.open }
+${ formattedTestResults }
+${ TEST_RESULTS_LIST.close }
+`;
+}
+
+function formatTestErrorMessage( flakyTestResult: FlakyTestResult ) {
+	switch ( flakyTestResult.runner ) {
+		case '@playwright/test': {
+			// Could do a slightly better formatting than this.
+			return stripAnsi(
+				flakyTestResult.results
+					.map( ( result ) => result!.error!.stack )
+					.join( '\n' )
+			);
+		}
+		case 'jest-circus':
+		default: {
+			return stripAnsi(
+				formatResultsErrors(
+					flakyTestResult.results,
+					{
+						rootDir: path.join(
+							process.cwd(),
+							'packages/e2e-tests'
+						),
+						// This is useless just to make typescript happy.
+						testMatch: [],
+					},
+					{ noStackTrace: false },
+					flakyTestResult.path
+				)!
+			);
+		}
+	}
+}
+
+function formatTestResults( {
+	date,
+	failedTimes,
+	headBranch,
+	runURL,
+	errorMessage,
+}: ParsedTestResult ) {
+	const dateString = date.toISOString();
+
+	// It will look something like this without formatting:
+	// [2021-08-31T16:15:19.875Z] Test passed after 2 failed attempts on trunk.
+	const log = `<time datetime="${ dateString }"><code>[${ dateString }]</code></time> Test passed after ${ failedTimes } failed ${
+		failedTimes === 1 ? 'attempt' : 'attempts'
+	} on <a href="${ runURL }"><code>${ headBranch }</code></a>.`;
+
+	if ( ! errorMessage ) {
+		return `${ TEST_RESULT.open }${ log }${ TEST_RESULT.close }`;
+	}
+
+	return `${ TEST_RESULT.open }<details>
+<summary>
+	${ log }
+</summary>
+
+\`\`\`
+${ errorMessage }
+\`\`\`
+</details>${ TEST_RESULT.close }`;
+}
+
+function parseFormattedTestResults(
+	formattedTestResults: string
+): {
+	date: Date;
+	failedTimes: number;
+	runURL: string;
+	headBranch: string;
+	errorMessage?: string;
+} {
```

Comment:
Can we add a `type` declaration for this? Just need to be careful about handling when new attributes are added.

Reply:
Sorry, so there's two comments here.

> Can we add a `type` declaration for this?

By this I just mean can we add a `type` alias for the `{ }` type that is returned by this function. Giving complex types a name improves readability in the same way that using variables improves readability. In particular I'm keen on anything that reduces the amount of "TypeScript noise" that can be distracting when reading code.

> Just need to be careful about handling when new attributes are added.

By this I mean we need to be careful that if, in the future, a new attribute is added to this type, that it is marked optional, because existing data that exists in GitHub will not have that new attribute. I'm not sure if there's a nice way to inform future maintainers of that—probably there's nothing we can do.

Reply:
Oh, it's the same type with `ParsedTestResult`, I totally forgot to change that 😅.